rep-0151.rst

```diff
@@ -0,0 +1,260 @@
+REP: 151
+Title: Python 3 Support
+Author: Dirk Thomas, Shane Loretz
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 05-Aug-2019
+
+Outline
+=======
+
+#. Abstract_
+#. Motivation_
+#. Context_
+#. Proposal_
+#. References_
+#. Copyright_
+
+Abstract
+========
+
+This REP describes the path for switching from Python 2 to Python 3 in ROS 1.
+
+Motivation
+==========
+
+As described in the Python release schedule [1]_ Python 2.7 will only receive
+bugfix releases until 2020.
+As a consequence Ubuntu is trying to demote Python 2.7 into the `universe`
+repository [2]_.
+While it is unlikely that Python 2.7 will be removed from the archives entirely,
+it should not be relied on past its EOL date.
+Therefore ROS 1 must move to Python 3.
+
+Context
+=======
+
+Different versions of Python can't be mixed within a single ROS distribution.
+A package using Python 3 can't import Python modules from a package built using
+Python 2.
```

Comment:
Packages built to support both Py2 and Py3 can import Python modules that have also been written to support both.  This is also a broadly proven fact even for numerous codebases that are more complex than ROS.

Reply:
If a package `foo` is installed with Python 2 its dependencies will use Python 2 packages which in the system are installed into a directory like `python2.7`. Another package `bar` installed with Python 3 would fail to use since dependencies of `foo` wouldn't necessarily be available for Python 3. Therefore we can't mix Python major versions within the same distro.

Reply:
You can but you would need to:
* __python2 -m pip install -r requirements.txt__
* __python3 -m pip install -r requirements.txt__

Doing so is such a common practice that saying it can not be done seems strange.

Reply:
You can absolutely do that when installing the package twice with `pip`. But that is not how a ROS distribution is built. Each ROS package is build into a Debian package. And in that process where we previously defaulted to Python 2 the upcoming ROS distro will default to Python 3.

Reply:
“_ROS Distributions are built by putting each dependency into a Debian package which prevents the simultaneous support of both Python 2 and Python 3._”

Reply:
We're currently doing our migration using the dual-python strategy proposed in #154. There have definitely been some sharp corners, particularly around feeding each interpreter the proper PYTHONPATH and ensuring correct shebangs in the installspace, but overall it has worked really well.

As in the proposal, we dual-build (via a catkin_tools plugin) any package containing a `setup.py` and ensure all system dependencies are available for both pythons, then migrate individual processes and scripts once they're ready and tested by changing their shebangs. This has allowed us to stage our migration over several months and avoid having to bring up our entire system again from scratch as Python 3.

So yes, not supporting dual-python workspaces upstream is a consequence of ROS packaging choices and ideology, not an inherent limitation in Python or anything else upstream.

Reply:
Sounds like you are on the right path.

Reply:
I deleted this paragraph in d48612e6414428242ae4cb2847247343ad9b48ee .

The Debian package context doesn't seem to help explain the rest of the document. Maybe a ROS distro could have targeted multiple Python versions by releasing Debian packages with a name prefix or suffix like upstream's `python-` or `python3-`, or maybe a ROS distro could have targeted multiple Python versions by installing duplicate dependencies for Python 2 and Python 3, but that's not what was chosen. Instead the context section starts with the link to the discourse post saying ROS Noetic will target Python 3 only.